spec/list_spec.rb

```diff
@@ -46,6 +46,12 @@ def fake_installed_xcodes(*names)
         installer.list.should == "1\n2.3\n2.3.1\n2.3.2\n3 some\n4 beta\n10 beta"
       end
 
+      it 'lists all versions filtered by requirements' do
+        fake_xcodes '1', '2.3', '2.3.1', '2.3.2', '3 some', '4 beta', '10 beta'
+        fake_installed_xcodes
+        installer.list('>= 2.0, < 10.0').should == "2.3\n2.3.1\n2.3.2\n3 some\n4 beta"
+      end
```

Comment:
This test is ok but it's quite simple. I'd be curious to know how Gem::Version treats more weird versions like "RC" "GM" "betas", etc. You could create multiple types of tests (<, >, <=, >=, ~>) for many different combinations from this list:

```
11
11.1
11.2
11.2.1
11.3 beta
11.3
11.3.1
11.4 beta
11.4 beta 2
11.4 beta 3
11.4
11.4.1
11.5 beta
11.5 beta 2
11.5 GM Seed
11.5
11.6 beta
11.6
11.7 beta
11.7
12 beta
12 beta 2
12 beta 3
12 beta 4
12 beta 5
12 beta 6
12 for macOS Universal Apps beta
12 for macOS Universal Apps beta 2
12
12.0.1
12.1.1 Release Candidate
12.1
12.2 beta
12.2 beta 2
12.2 beta 3
12.2 Release Candidate
12.2
12.3 beta
12.3 Release Candidate
12.3
12.4 Release Candidate
12.4
12.5 beta
12.5 beta 2
12.5 beta 3
12.5 Release Candidate
12.5
12.5.1
13 beta
13 beta 2
13 beta 3
13 beta 4
13 beta 5
13
13.1
13.2 beta
13.2 beta 2
13.2 Release Candidate
13.2
13.2.1
13.3 beta
13.3 beta 2
13.3 beta 3
13.3 Release Candidate
13.3 (installed)
13.3.1
13.4 Release Candidate
13.4
13.4.1
14 beta
14 beta 2
```

And see what you get. Thoughts? Just trying to make sure this code will actually work "as expected" for this particular use case. Not even sure what's truly "expected" as a user, for those non-standard version names 🤔

Reply:
Agreed, it's not straightforward to get a sense of the expected or correct behaviour of the `list --filter` command. We are essentially comparing two different things here: one is Xcode version, which isn't a semantic version, and the other is the `filter` option value, which is a semantic versioning requirements expressed by [`Gem::Requirement`](https://ruby-doc.org/stdlib-2.6.4/libdoc/rubygems/rdoc/Gem/Requirement.html).

This PR was made quite a while ago, now I'm looking back at the implementation, I don't feel like it produces an easy-to-understand behaviour.

The current implementation treats all Xcode releases under the same version, including beta and RC, as the normal version. This means, when passing `--filter='> 13.2-beta'` option intending to list all versions released after the first beta 13.2, all 13.2 versions, including "13.2 Beta", will be printed out.

One easy improvement I can think of is, translating the Xcode version to semantic version, by simply using a `-` to concatenate the normal version and the pre-release version when exists. For example, "13.2 Beta 2" becomes "13.2-Beta.2", "13.3 Release Candidate" becomes "13.3-Release.Candidate", and "12 for macOS Universal Apps beta 2" becomes "12-for.macOS.Universal.Apps.beta.2" (which looks pretty weird). The translated version string can be used to create a `Gem::Version`, for checking if it satisfies the requirements specified by the `filter` option. This would fix the undesired behaviour mentioned above.

Similar improvement can also be performed on the `filter` option, whose value is used to create a `Gem::Requirement`. This means users must use semantic versions, _not_ Xcode versions, otherwise an exception will occur. You've seen an example of this above: note the version in this command line option, `--filter='> 13.2-beta'`, "13.2-beta" is used here, not "13.2 Beta". But `> 13.2 Beta` probably would be the first thing that came to users' minds, since this tool is dealing with Xcode versions after all. We can perform the same translation mentioned above on this option value so that Xcode versions become acceptable.

Let me know what you think.

Reply:
Hi @crazytonyli , thanks for getting back to this with such thorough response!

Taking a step back and comparing the two options, both has its maintenance drawbacks. However, I think that if we can cleanly and somewhat reliably automatically convert Apple's weird Xcode version naming conventions to a true semantic versioning naming convention, I think that is the option we can benefit most from. The use case possibilities expand:

- Other features similar features could be built around this new functionality
- The Xcode file installed in `/Applications` could then be named after this, which makes more sense and could give us exact matches more easily
- The `.xcode-version` files could then be expressed using semantic versioning, which I believe it makes more sense to readers in general, to follow one universal standard. And this doesn't mean we'd lose readability or anything (it's not like the avg person wouldn't understand the now-program-readable version string).

With that being said, if we did this, it would obviously break all existing setups so we'd need a major bump (xcode-install v3.0.0).

After exposing these benefits, do you see other drawbacks, challenges or other considerations we should take here? Future compatibility issues with Xcode versions yet to be released? I guess I don't fully understand those less-known areas of semantic versioning (other than the obv x.y.z standard), so I'm trust your understanding 😊 but I reckon I need to read up on it before we commit to this.

Separate matter: would you be willing to take on this effort? 🙇

Reply:
Interesting, that does sound like a significant change. My preference would be to start with making it a hidden implementation detail. If it works well, with a few future Xcode releases, maybe consider making it a "public API" then.

Regarding using the converted semantic version to replace the official Xcode version, one concern I have is how straightforward will this conversion logic be for use to get, without the need of reading any documentation or (in the worst case scenario) the ruby implementation. Here are a few questions I can think of (but don't have good answers for):
- Should the semantic version of "Xcode 14 beta" be 14.0.0-beta, or 14.0.0-beta.1?
- If we go with "14.0.0-beta", what if Apple releases the first Xcode 15 beta as "Xcode 15 beta 1"? Do we use then follow Apple by using "beta.1" for Xcode 15, but not for 14?
- Do we omit the minor.patch part if they are "0", like how Apple does?

To be honest, I'm not confident in making this name conversion intuitive enough as an easy-to-understand "public API". Of course, none of the above questions matters if this conversion is an implementation hidden from users.

Reply:
What do you mean exactly by making it hidden from the users?

Reply:
> if this conversion is an implementation hidden from users.

I was referring to [the suggestion I made above](https://github.com/xcpretty/xcode-install/pull/438#discussion_r912461817). Where users use Xcode versions in `xcversion list --filter=...`, with the expectation that the command produces results based on semantic version.

Reply:
Gotcha, yeah that could work! :)

Reply:
Please see latest changes I just pushed up. I hope all updated behaviours are covered by the new unit tests. Let me know what you think, thanks! 😺